packages/reqresp/src/rate_limiter/ReqRespRateLimiter.ts

```diff
@@ -0,0 +1,153 @@
+  private readonly rateLimitersTotal = new Map<ProtocolMethod, Map<ProtocolVersion, RateLimiterGRCA<null>>>();
+  /** Interval to check lastSeenMessagesByPeer */
+  private cleanupInterval: NodeJS.Timeout | undefined = undefined;
+  private rateLimitMultiplier: number;
+  /** Periodically check this to remove tracker of disconnected peers */
+  private lastSeenRequestsByPeer: Map<string, number>;
+  private readonly metrics: Metrics | null;
+  private readonly logger: ILogger;
+  private readonly reportPeer: ReqRespRateLimiterModules["reportPeer"];
+
+  constructor({metrics, logger, reportPeer}: ReqRespRateLimiterModules, {rateLimitMultiplier}: ReqRespRateLimiterOpts) {
+    this.metrics = metrics;
+    this.reportPeer = reportPeer;
+    this.logger = logger;
+    this.rateLimitMultiplier = rateLimitMultiplier ?? 1;
+    this.lastSeenRequestsByPeer = new Map();
+  }
+
+  get enabled(): boolean {
+    return this.rateLimitMultiplier > 0;
+  }
+
+  initRateLimits<Req, Resp>(protocol: ProtocolDefinition<Req, Resp>): void {
+    if (!this.enabled) {
+      return;
+    }
+
+    const {method, version} = protocol;
+    const limiterByPeer = RateLimiterGRCA.fromQuota<string>({
+      ...protocol.inboundRateLimits.byPeer,
+      quota: protocol.inboundRateLimits.byPeer.quota * this.rateLimitMultiplier,
+    });
+    const limiterTotal = RateLimiterGRCA.fromQuota<null>({
+      ...protocol.inboundRateLimits.total,
+      quota: protocol.inboundRateLimits.total.quota * this.rateLimitMultiplier,
+    });
+
+    if (!this.rateLimitersPerPeer.has(method)) {
+      this.rateLimitersPerPeer.set(method, new Map());
+    }
+
+    if (!this.rateLimitersTotal.has(method)) {
+      this.rateLimitersTotal.set(method, new Map());
+    }
+
+    this.rateLimitersPerPeer.get(method)?.set(version, limiterByPeer);
+    this.rateLimitersTotal.get(method)?.set(version, limiterTotal);
+  }
+
+  validateRateLimits<Req, Resp>({
+    peerId,
+    protocol,
+    requestBody,
+  }: {
+    peerId: PeerId;
+    protocol: ProtocolDefinition<Req, Resp>;
+    requestBody: Req;
+  }): void {
+    if (!this.enabled) {
+      return;
+    }
+    const peerIdStr = peerId.toString();
+    this.lastSeenRequestsByPeer.set(peerIdStr, Date.now());
+    const {method, version, encoding} = protocol;
+    const requestCount = protocol.inboundRateLimits.getRequestCount
+      ? protocol.inboundRateLimits.getRequestCount(requestBody)
+      : 1;
+
+    if (!this.rateLimitersPerPeer.get(method)?.get(version)?.allows(peerIdStr, requestCount)) {
+      this.logger.debug("Do not serve request due to rate limit", {
+        peerId: peerIdStr,
+      });
+
+      this.reportPeer(peerId);
+
+      if (this.metrics) {
+        this.metrics.rateLimitErrors.inc({tracker: "requestCountPeerTracker"});
+      }
+
+      throw new RequestError(
+        {
+          code: RequestErrorCode.REQUEST_RATE_LIMITED,
```

Comment:
Remove new line after `{` for consistency with rest of codebase. Same for `{
        peerId: peerIdStr,
      }`

Reply:
I formatted this manually for now, but such things should be left to the `prettier` and should not be engaged in the code review cycle.